scripts/admin/cmerge

```diff
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+# @author Dominic Jäger <e1634025@student.tuwien.ac.at>
+# @brief Perform a three-way merge with files using Elektra
+# @date 2019-08-20
+# @tags creator
+
+ourpath="user/elektra-cmerge/our"
+theirpath="user/elektra-cmerge/their"
+basepath="user/elektra-cmerge/base"
+resultpath="user/elektra-cmerge/result"
+resultfile="elektra-cmerge_result"
+
+if [ "$#" -ne 4 ]; then
+	echo "Usage is $0 our their base format"
+	exit 1
+fi
+format=$4
+kdb rm -rf $ourpath
```

Comment:
Better to avoid this: make an import strategy that overwrites all existing keys.

Reply:
> make an import strategy

That means make a strategy for `kdb import`, not make a strategy called `import` (for the new merge), right?

Currently, our use case is actually fulfilled and my `kdb rm -rf` seems to have been redundant.
```
jenkins@43c319066fd2:~/workspace/elektra-build-docker$ echo -e "anton=a\nberta=a" | kdb import user ini
jenkins@43c319066fd2:~/workspace/elektra-build-docker$ kdb ls user
user/anton
user/berta
jenkins@43c319066fd2:~/workspace/elektra-build-docker$ echo -e "xaver=x" | kdb import user ini
jenkins@43c319066fd2:~/workspace/elektra-build-docker$ kdb ls user
user/xaver
```


For the strategy you mentioned
```c
diff --git a/src/tools/kdb/import.cpp b/src/tools/kdb/import.cpp
index a56100076..45446ffdd 100644
--- a/src/tools/kdb/import.cpp
+++ b/src/tools/kdb/import.cpp
@@ -99,6 +99,17 @@ int ImportCommand::execute (Cmdline const & cl)
 		KeySet systemKeySet = baseCopy.cut (systemElektra);
 		importedKeys.append (systemKeySet);
 	}
+	/** This option removes all keys that existed at the path before.
+	 * It is equal to calling kdb rm -rf <path> before the import
+	 **/
+	if (cl.strategy == "remove")
+	{
+		if (ckdb::ksCopy (base.release (), NULL) != 0)
+		{
+			throw "Could not remove keys during import";
+		}
+	}
 
 	ThreeWayMerge merger;
 	MergeHelper helper;
```

but it won't really work due to the calls for the C++ merge around it. For example, I cannot specify the strategy "remove" on the command line as it is not an accepted strategy by the C++ merge. In some way it would be useful to throw the whole C++ merge out of import to create the new strategy. What do you think, @markus2330 ?

Reply:
> That means make a strategy for kdb import, not make a strategy called import (for the new merge), right?

Strategies should be available for every tool (as they are a feature of the elektraMerge API). Maybe start with a document describing the strategies (doc/help/elektra-merge-strategy.md) so that we get good and consistent names for the strategies.

> What do you think

Yes, all this code should be removed and instead a single elektraMerge call should suffice.

This would make it much easier to maintain the same behavior across tools.

Reply:
> Strategies should be available for every tool (as they are a feature of the elektraMerge API)

This makes sense to me and this is why there is a parameter "strategy" in this API call. However, the call currently takes three KeySets. I thought of it as you take 3 sets (wherever they come from) and get a single one as result. And then then strategy decides what to do with those sets.

Following this thought, removing everything that has been at the _path_ is not part of the API. Something would have to clear everything from the path. Afterwards, we could create a set from this path and put it into the API.

An import strategy would mean that you put any sets into the API and say "Please remove all the keys in them before doing anything else". Only looking at the API, this seems strange to me as we could have just used an empty set to begin with.


What we could do is 
1. change this idea of the API completely (seems bad... all the unit tests 😟)
2. write a wrapper around elektraMerge (elektraMergePaths maybe?). Actually the `kdb` tool does exactly this. It gets three paths and calls the API.

> Maybe start with a document describing the strategies

They are currently in the cmerge man page, but I can also make a separate document of this.

> Yes, all this code should be removed and instead a single elektraMerge call should suffice.

Now already? Not later when we drop the C++ merge completely?

Reply:
Yes I agree, there is also the problem that the flags "that modify the result set" and the "strategy" flags can be combined in any way. So yes, `elektraMergeResult(ours, theirs, base, char * strategy, result, char * behavior)` might be the solution for tooling (and still providing a normal 3-way merge, called elektraMerge, for non-tools, i.e., that do not write the result to KDB).

This also means we do not only have a strategy, but also the behavior for what to do with the result (e.g. to wipe out, to overwrite, to fail on overwrite, to fail on non-empty).

> They are currently in the cmerge man page, but I can also make a separate document of this.

Yes, it is better separated as several tools have the same strategies/overwrite behaviors.

> Now already? Not later when we drop the C++ merge completely?

At least for web-ui we should do it now (it did not have merging before and C++ would also be difficult to be used in Go). Then we would finally see merging in action.

Reply:
Sounds good. Would it be OK for you if I make this as new PR? It would be great to be able to merge this PR asap as keeping track of the different branches is bit exhausting. For example,  #3128 and #3141 are based on some commit of this PR and I have some more (backup) branches locally.